src/ripple/nodestore/impl/Database.cpp

```diff
@@ -0,0 +1,253 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/nodestore/Database.h>
+#include <ripple/basics/chrono.h>
+#include <ripple/beast/core/CurrentThreadName.h>
+#include <ripple/protocol/HashPrefix.h>
+
+namespace ripple {
+namespace NodeStore {
+
+Database::Database(std::string name, Stoppable& parent,
+    Scheduler& scheduler, int readThreads, beast::Journal journal)
+    : Stoppable(name, parent)
+    , j_(journal)
+    , scheduler_(scheduler)
+{
+    while (readThreads-- > 0)
+        readThreads_.emplace_back(&Database::threadEntry, this);
+}
+
+Database::~Database()
+{
+    // NOTE!
+    // Any derived class should call the stopThreads() method in its
+    // destructor.  Otherwise, occasionally, the derived class may
+    // crash during shutdown when its members are accessed by one of
+    // these threads after the derived class is destroyed but before
+    // this base class is destroyed.
+    stopThreads();
+}
+
+void
+Database::waitReads()
+{
+    std::unique_lock<std::mutex> l(readLock_);
+    // Wake in two generations
```

Comment:
The code and comment come from the previous DatabaseImp class. I have no idea why 2 was chosen.

Reply:
Each generation is a full pass over the space. If we're in generation N and you issue a request, that request will only be done during generation N if it happens to land after where the pass currently is. But it will definitely be done during generation N+1 since the request was in the table before that pass even started. So when you reach generation N+2, you know the request is done.

Reply:
Thanks @JoelKatz , I will document this in the code.